File: leetcode/1110_delete_nodes_return_forest.cpp


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def delNodes(self, root: TreeNode, to_delete: List[int]) -> List[TreeNode]:
        to_delete_set = set()
        for node in to_delete:
            to_delete_set.add(node)
        result = []
        self.pre_order(result, to_delete_set, root, True)
        return result

    def pre_order(self, result, to_delete_set, node, is_root):
        if node == None:
            return None
        is_deleted = node.val in to_delete_set
        if (is_root and not is_deleted):
            result.append(node)
        node.left = self.pre_order(result, to_delete_set, node.left, is_deleted)
        node.right = self.pre_order(result, to_delete_set, node.right, is_deleted)
        return None if is_deleted else node

File: leetcode/1110_delete_nodes_return_forest.py
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        vector<TreeNode*> result;
        set<int> to_delete_set;
        vector<int>::iterator itr;
        for(itr=to_delete.begin(); itr!=to_delete.end(); itr++) {
            to_delete_set.insert(*itr);
        }
        preOrder(root, result, to_delete_set, true);
        return result;
    }
    TreeNode* preOrder(TreeNode* node, vector<TreeNode*>& result, set<int>& to_delete_set, bool is_root) {
        if (node == NULL) {
            return NULL;
        }
        bool deleted = to_delete_set.find(node->val) != to_delete_set.end();
        if (is_root && !deleted) {
            result.push_back(node);
        }
        node->left = preOrder(node->left, result, to_delete_set, deleted); // using the fact that when a node is deleted, it's left and right children forms two different trees in the forest
        node->right = preOrder(node->right, result, to_delete_set, deleted);
        return deleted ? NULL : node;
    }
};
